wireapp/kalium
# feat: discover legal hold when receiving message [WPB-5998]

----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

When a conversation starts to become under legal hold, we need to discover that on Android when receiving messages and handle this change properly by showing system messages.

### Solutions

Implement new function inside `LegalHoldHandler` to handle legal hold with new messages. 
Rearrange legal hold handlers so that `LegalHoldSystemMessagesHandler` is indeed responsible only for system messages and doesn't change any legal hold state.
According to the [documentation](https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/157255445/Use+Case+Discover+legal+hold+on+receiving+a+message), when syncing, messages are buffered and handled together after the sync by using new `TriggerBuffer` which works kind of like a valve - it takes a `Boolean` flow and emits items while it's true.

### Testing

#### Test Coverage (Optional)


#### How to Test

Enable legal hold for some other member before logging in and syncing, then login - at this point the app doesn't yet know that a conversation or member is under legal hold, but after receiving first message, it should be handled properly.

### Attachments (Optional)

https://github.com/wireapp/kalium/assets/30429749/5c33b200-5b6f-4b7c-8913-27d611df4b11

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
